File: 05_Graph_Search/Level4/1753.java
// 최단경로
// https://www.acmicpc.net/problem/1753
// 힌트
// 1. BFS알고리즘을 이용하면 메모리 초과가 발생한다.
// 2. Dijkstra알고리즘을 이용하여 시작점으로 각 노드에 대한 최단 거리를 구한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Edge implements Comparable<Edge>{
	int to;
	int weight;
	
	public Edge(int to, int weight) {
		this.to = to;
		this.weight = weight;
	}

	@Override
	public int compareTo(Edge o) {
		return this.weight - o.weight;
	}
}

public class Main {
	static int V;
	static ArrayList<ArrayList<Edge>> edges;
	static int minDist[];
	
	static void dijkstra(int start) {
		PriorityQueue<Edge> q = new PriorityQueue<>();
		
		q.add(new Edge(start, 0));
		
		while(!q.isEmpty()) {
			Edge ed = q.poll();
			int s = ed.to;
			for (Edge e: edges.get(s)) {
				int to = e.to;
				int weight = e.weight;
				if (minDist[s] + weight < minDist[to]) {
					minDist[to] = minDist[s] + weight;
					q.add(new Edge(to, minDist[to]));
				}
				
			}
		}
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		V = Integer.parseInt(st.nextToken());
		int E = Integer.parseInt(st.nextToken());
		int K = Integer.parseInt(br.readLine());
		
		edges = new ArrayList<>();
		for (int i = 0; i < V + 1; i++) {
			edges.add(new ArrayList<Edge>());
		}
		minDist = new int[V + 1];
		Arrays.fill(minDist, Integer.MAX_VALUE);
		minDist[K] = 0;
		
		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int u = Integer.parseInt(st.nextToken()); 
			int v = Integer.parseInt(st.nextToken());
			int w = Integer.parseInt(st.nextToken());
			edges.get(u).add(new Edge(v, w));
		}
		
		dijkstra(K);
		
		for (int i = 1; i < V + 1; i++) {
			if (minDist[i] == Integer.MAX_VALUE) {
				System.out.println("INF");
			} else {
				System.out.println(minDist[i]);
			}
		}
	}
}

File: 05_Graph_Search/Level4/1753.py
# 최단경로
# https://www.acmicpc.net/problem/1753
# 힌트
# 1. BFS알고리즘을 이용하면 메모리 초과가 발생한다.
# 2. Dijkstra알고리즘을 이용하여 시작점으로 각 노드에 대한 최단 거리를 구한다.
import sys
import heapq


def dijkstra():
    hq = []

    heapq.heappush(hq, (0, K))

    while len(hq) > 0:
        w, v = heapq.heappop(hq)

        for edge in edges[v]:
            vw, vt = edge
            if min_dist[vt] > min_dist[v] + vw:
                min_dist[vt] = min_dist[v] + vw
                heapq.heappush(hq, (min_dist[vt], vt))


if __name__ == "__main__":
    V, E = map(int, sys.stdin.readline().split())
    K = int(sys.stdin.readline())

    min_dist = [sys.maxsize] * (V+1)
    min_dist[K] = 0

    edges = [[] for _ in range(V + 1)]

    for _ in range(E):
        u, v, w = map(int, sys.stdin.readline().split())
        edges[u].append((w, v))

    dijkstra()

    for i in range(1, V + 1):
        if min_dist[i] == sys.maxsize:
            print('INF')
        else:
            print(min_dist[i])